sds: skip waiting for ingress gateway secret for file mounted certs

ramaraochavali opened the pull request:
Continuing on #23647 , this PR skips waiting for ingress gateway secret for file mounted certs

ramaraochavali commented:
@lei-tang @myidpt Can you PTAL? This is also need to be skipped for file mounted certs

howardjohn commented on security/pkg/nodeagent/sds/sdsservice_test.go:
```diff
@@ -1023,7 +1023,9 @@ func (ms *mockSecretStore) DeleteSecret(conID, resourceName string) {
-	return false
+	// Return true here for file mounted cert resource name to validate that file mounted certs would
+	// not depend on this result and proceed with generating secret.
+	return strings.HasPrefix(resourceName, "file-root:")
```
Is this implemented for Gateway? I thought it was just DestinationRule

ramaraochavali replied:
It was just implemented for DestinationRule and other server certs. But the sds service https://github.com/istio/istio/blob/master/security/pkg/nodeagent/sds/sdsservice.go#L328 here waits for it

lei-tang commented on security/pkg/nodeagent/sds/sdsservice_test.go:
```diff
@@ -1023,7 +1023,9 @@ func (ms *mockSecretStore) DeleteSecret(conID, resourceName string) {
 
 func (ms *mockSecretStore) ShouldWaitForIngressGatewaySecret(connectionID, resourceName, token string) bool {
-	return false
+	// Return true here for file mounted cert resource name to validate that file mounted certs would
```
I am unsure that the comment added here is accurate: "// Return true here for file mounted cert resource name to validate that file mounted certs would not depend on this result and proceed with generating secret.".
1. What result would not be depended?
2. Can you revise the aforementioned comment to make its meaning easier to understand (e.g., break this long sentence into multiple shorter sentences; and explain 1) at which scenario which resources will be waited, 2) at which scenarios which resources will not be waited)?

ramaraochavali replied:
Is this comment clear now? or do you want me to add any thing more?

lei-tang replied:
According to the code, when  strings.HasPrefix(resourceName, "file-root:") returns true, ShouldWaitForIngressGatewaySecret() returns true, which means waiting for ingress gateway secret.
Can you explain why waiting for ingress gateway secret or not is determined by the "file-root" resource?

ramaraochavali replied:
With the new change this check not needed because it moves to ShouldWaitForIngressGatewaySecret. I will remove it

lei-tang commented on security/pkg/nodeagent/sds/sdsservice.go:
```diff
@@ -325,9 +328,13 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
-			if s.st.ShouldWaitForIngressGatewaySecret(conID, resourceName, token) {
-				sdsServiceLog.Warnf("%s waiting for ingress gateway secret for proxy %q\n", conIDresourceNamePrefix, discReq.Node.Id)
-				continue
+			if !s.fileMountedCerts {
```
1. To improve code readability, can you add a comment to explain the difference between the file mounted cert and the ingress gateway secret? 
2. To improve code readability, can you add a comment to explain whether ingress gateway secret is also mounted liked the file mounted cert or ingress gateway secret is provisioned through a different mechanism?

ramaraochavali replied:
@lei-tang do you mean this comment? I have updated it. Could you please let me know if I am missing any thing?

lei-tang replied:
The missing items are:
1) Can you add a comment to explain the difference between the file mounted cert and the ingress gateway secret?
2) Can you add a comment to explain whether ingress gateway secret is also mounted liked the file mounted cert or ingress gateway secret is provisioned through a different mechanism?

ramaraochavali replied:
It is there in the comment already, I tweaked the comment a little bit to make it more clear. Let me know if you think any thing else needed to be clarified.

ramaraochavali commented:
@lei-tang added more comments. PTAL.

lei-tang commented:
@ramaraochavali I have read the newly added comments in the code. Can you also address the review comments I added?

lei-tang commented on security/pkg/nodeagent/sds/sdsservice.go:
```diff
@@ -325,9 +328,15 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
-				continue
+			// If workload uses file mounted certs i.e. "FILE_MOUNTED" is set to true, it does not depend on ingress gateway
+			// secret setup as a kubernetes secret - we should skip waiting for it in that mode.
+			if !s.fileMountedCerts {
```
To simplify the code and improve the code readability, s.fileMountedCerts can be incorporated into ShouldWaitForIngressGatewaySecret(). The simplfied code will be like the following:

if s.st.ShouldWaitForIngressGatewaySecret(conID, resourceName, token, s.fileMountedCerts) {
					sdsServiceLog.Warnf("%s waiting for ingress gateway secret for proxy %q\n", conIDresourceNamePrefix, discReq.Node.Id)
					continue
}

ramaraochavali replied:
Changed it like this and added a specific test for filemounted cert. PTAL

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 2af63ac214d1ae5ba68eb5adce551b80187f8906 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23986/unit-tests_istio/13678) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test unit-tests_istio

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-telemetry-k8s-tests_istio | 2af63ac214d1ae5ba68eb5adce551b80187f8906 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23986/integ-telemetry-k8s-tests_istio/13159) | `/test integ-telemetry-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-telemetry-k8s-tests_istio

ramaraochavali commented:
@lei-tang Can you PTAL when you get chance?

ramaraochavali commented:
@myidpt I have addressed all the comments of @lei-tang . Could you please take a look? I will address if @lei-tang has further comments in a follow-up.

lei-tang commented:
@ramaraochavali Thanks for the revisions, with which the PR is in a much better shape. 

1. Can you point me to the place where you explain the difference between the file mounted cert and the ingress gateway secret? Without an explanation of the difference between the file mounted cert and the ingress gateway secret, a user will have difficulties to use the feature of the file mounted cert.

2. Can you point me to the place where you explain whether ingress gateway secret is also mounted liked the file mounted cert or ingress gateway secret is provisioned through a different mechanism?

ramaraochavali commented on security/pkg/nodeagent/sds/sdsservice.go:
```diff
@@ -322,12 +325,16 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
-			// for a new kubernetes secret before sending SDS response.
-			if s.st.ShouldWaitForIngressGatewaySecret(conID, resourceName, token) {
+			// In ingress gateway agent mode, if the first SDS request is received but Ingress gateway secret which is
+			// setup as kubernetes secret is not ready, wait for secret before sending SDS response.
```
@lei-tang In this comment
"// In ingress gateway agent mode, if the first SDS request is received but Ingress gateway secret which is
// setup as kubernetes secret is not ready, wait for secret before sending SDS response."

I explained how Ingress gateway secret is setup i.e. as kubernetes secret
In the below I explained how file mounted certs are setup with the env flag.

If you are looking for any this else to be documented - please suggest. I will add

lei-tang replied:
With the existing ingress gateway secret mechanism, why the mechanism of file mounted certs is needed for ingress gateway?

ramaraochavali replied:
I just added some more explanation. PTAL. Please suggest if if you have other points. Will add

lei-tang replied:
I am not sure which explanation in your PR answers the question: with the existing ingress gateway secret mechanism, why the mechanism of file mounted certs is needed for ingress gateway?

To clearly anser the question, please paste your answer to the above question as a comment.

linsun commented on security/pkg/nodeagent/sds/sdsservice.go:
```diff
@@ -322,12 +325,16 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
+			// In ingress gateway agent mode, if the first SDS request is received but Ingress gateway secret which is
+			// setup as kubernetes secret is not ready, wait for secret before sending SDS response.
+			// If a kubernetes secret was deleted by operator, wait for a new kubernetes secret before sending SDS response.
+			// If workload uses file mounted certs i.e. "FILE_MOUNTED" is set to true, it does not depend on ingress gateway
```
Hi @ramaraochavali - Where is "FILE_MOUNTED" set to true?

linsun replied:
BTW I assume we support both (file mounted and sds gateway) ?  and depends on how secrets/credentials are specified within the user's gateway resource, the code determines which path it should be used.

cc @howardjohn

ramaraochavali replied:
https://github.com/istio/istio/blob/7cb751f183999845ad58100589584c4084838463/pkg/istio-agent/sds-agent.go#L89 - It was added a while ago to pilot-agent (specically to sds-agent) to indicate a workload is using file mounted certs. Deployments which have that can set that env var

linsun replied:
Thank you @ramaraochavali !  I think this may be @lei-tang 's question as well.

lei-tang replied:
With the existing ingress gateway secret mechanism, why the mechanism of file mounted certs is needed for ingress gateway? What is the pros and cons of the existing ingress gateway secret mechanism versus the proposed file-mounted certs for ingress gateway?

howardjohn replied:
People have existing CAs/secret management that inject certs as files in the pod. Also VMs...

For the average case you would always use Secret reading - but if you have existing integrations it's simply not possible

ramaraochavali replied:
Correct. And additionally for sidecar with has `FILE_MOUNTED_CERTS` to true, this fails to loads those certs and send to proxy with this error
```
2020-05-20T10:14:43.765581Z	warn	sds	resource:file-root:/etc/identity/ca/cacerts.pem waiting for ingress gateway secret for proxy "sidecar~10.100.172.250~istio-ordering-98c49bb4c-q854l.service-mesh~service-mesh.svc.cluster.local"
```

lei-tang replied:
For non-ingress workloads, the mechanism of file mounted certs makes sense because no equivalent mechanisms exist for non-ingress workloads.
But for ingress gateways, the existing ingress gateway secret mechanism is equivalent to the proposed mechanism of file mounted certs. Adding an equivalent mechanism to the ingress gateway will make the code more complicated and may create confusions to the users.

howardjohn replied:
Lei, there are many many tools that integrate to provision certificate into pod. https://www.hashicorp.com/blog/injecting-vault-secrets-into-kubernetes-pods-via-a-sidecar/ is one example. I don't think it's reasonable to ask people to change their entire PKI infrastructure

ramaraochavali replied:
> But for ingress gateways, the existing ingress gateway secret mechanism is equivalent to the proposed mechanism of file mounted certs
It is actually not equivalent. Because @howardjohn  mentioned, in Salesforce (and I am sure lot of other enterprises will also have similar setup) we have an existing PKI infrastructure that manages cert delivery and rotation as file in PODS (for non-ingreess as well as ingress) which is security vetted.  I think we can not assume kubernetes secrets.,

Also apart from Ingress, this check/wait is also blocking sidecars to come up. What about VMs? What about deployments that do not use ingress gateway? In those cases this should not block sidecars to come up

lei-tang replied:
Adding more mechanisms to provision certificates into pods may help users. But to make the proposed mechanism truly useful to potential users, the PR should help users to understand the proposed mechanism, e.g., describing the differences between the proposed mechanism and the existing mechanisms, as well as their different usage scenarios. The review comments have repeatedly asked for such information and finally we see some answers in the comments.

ramaraochavali replied:
Sorry, it was not completely clear to me. I have added comment about usage scenario for file mounted certs. But I am happy to additional comments if you have any suggestions

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | ce2610c1de10dab5ea3baf73e65750ade7c91c9d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23986/lint_istio/13207) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 7b97d15a2fcab0b57549d68bf922a2ac9188ca57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23986/lint_istio/13208) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

lei-tang commented on security/pkg/nodeagent/sds/sdsservice.go:
```diff
@@ -331,6 +331,8 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
 			// If workload uses file mounted certs i.e. "FILE_MOUNTED_CERTS" is set to true, workdload loads certificates from
 			// mounted certificate paths and it does not depend on the presence of ingress gateway secret so
 			// we should skip waiting for it in that mode.
+			// File Mounted certs for gateways is used in scenarios where an existing PKI infrastuctures delivers certificates
```
nit: "File Mounted" -> "File mounted".

ramaraochavali replied:
changed

ramaraochavali commented:
Thank you @lei-tang . @myidpt  Can you PTAL when you get chance?

myidpt commented on security/pkg/nodeagent/sds/sdsservice.go:
```diff
@@ -322,12 +325,19 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
 				sdsServiceLog.Warnf("%s waiting for ingress gateway secret for proxy %q\n", conIDresourceNamePrefix, discReq.Node.Id)
 				continue
+			} else {
+				sdsServiceLog.Infof("Workload is using file mounted certificates. Skipping waiting for ingress gateway secret")
```
This log is not correct. There can be multiple reasons ShouldWaitForIngressGatewaySecret returns false.

ramaraochavali replied:
That is a great point. I will change the log

myidpt commented on security/pkg/nodeagent/sds/sdsservice.go:
```diff
@@ -322,12 +325,19 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
+			// we should skip waiting for it in that mode.
+			// File mounted certs for gateways is used in scenarios where an existing PKI infrastuctures delivers certificates
+			// to pods/VMs via files.
+			if s.st.ShouldWaitForIngressGatewaySecret(conID, resourceName, token, s.fileMountedCerts) {
```
Do we need to pass fileMountedCerts into ShouldWaitForIngressGatewaySecret? Is it better to check the condition here?

ramaraochavali replied:
This was changed as per previous https://github.com/istio/istio/pull/23986#discussion_r428420884

myidpt commented on security/pkg/nodeagent/sds/sdsservice.go:
```diff
@@ -146,7 +148,7 @@ type Debug struct {
 
 // newSDSService creates Secret Discovery Service which implements envoy v2 SDS API.
-func newSDSService(st cache.SecretManager, skipTokenVerification, localJWT bool,
+func newSDSService(st cache.SecretManager, skipTokenVerification, localJWT, fileMountedCerts bool,
```
Ingress gateway can support fetching from secret + file mount, right? @JimmyCYJ 
This flag means only file mount is used? If so, we should make it explicit in the flag. Using a name like "fileMountedCertsOnly".

ramaraochavali replied:
I am not sure if such a case exists. But there is no harm in changing the var name. So I changed

ramaraochavali commented:
@myidpt Changed. PTAL

ramaraochavali commented:
@myidpt Could you PTAL when you get chance?

ramaraochavali commented:
Thank you @myidpt

istio-testing commented:
In response to a cherrypick label: new pull request created: #24254
